### PR TITLE
Display the component options to a NoSuperUser account

### PR DESCRIPTION
**Pull request by carlitorweb**
Pull Request for Issue #39706 

`Joomla\CMS\Dispatcher\ComponentDispatcherFactory::createDispatcher` search for the component dispatcher, in this case the com_config dispatcher. As there is not a dispatcher class inside the component, is used `Joomla\CMS\Dispatcher\ComponentDispatcher::checkAccess()`. This function check only if the user have `core.manage` permission, something com_config do not need kown from a NoSuperUser.

**One solution is change `Joomla\CMS\Dispatcher\ComponentDispatcher::checkAccess()` for check in case:**

```php
$this->option === "com_config"
``` 

then check for the right permission:

```php
!$this->app->getIdentity()->authorise('core.options', $this->option)
```

But I liked better the approch of override the function in the component.

### Testing Instructions
Check the issue description

### Actual result BEFORE applying this Pull Request
Check the issue description

### Expected result AFTER applying this Pull Request
Clicking on the Options button will display the options for the component.

### Link to documentations
Please select:
- [x] No documentation changes for docs.joomla.org needed

- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by carlitorweb** [comment]
For maintainers, please notice `checkAccess()` can be empty. I added that check there for be sure we cover all possible scenarios. Let me kown if keep the function empty is good enough.

**Comment by wilsonge** [comment]
Pinging @joomla/security

**Comment by joomdonation** [comment]
I looked at com_config and saw that the code provided in this PR does not cover all the cases yet. I suggest use the code like below:

```php
protected function checkAccess()
{
	$task = $this->input->getCmd('task');

	// sendtestmail expects json response, so we leave the method to handle the permission and send response itself
	if ($task === 'application.sendtestmail') {
		return;
	}

	$option = $this->input->getCmd('component');
	$user   = $this->app->getIdentity();

	if ($option && !in_array(strtolower($option), ['com_joomlaupdate', 'com_privacy'], true)) {
		$canAccess = $user->authorise('core.admin', $option) || $user->authorise('core.options', $option);
	} else {
		$canAccess = $user->authorise('core.admin');
	}

	if (!$canAccess) {
		throw new NotAllowed($this->app->getLanguage()->_('JERROR_ALERTNOAUTHOR'), 403);
	}
}
```

Explanation:

-  The checkAccess method should throw `NotAllowed` exception if user does not have permission to perform the action like we do for all other extensions (not perform redirection like in proposed code)
-  When component is provided in the request, we check for `core.admin` or `core.options` permissions for all components except **com_privacy**, **com_joomlaupdate** . For the two components, we need to have global **core.admin** permission.
- For all other actions in com_config, users need to have global **core.admin** permission.
- We need special case for `sendtestmail` here because the ajax request expect json response, so if we throws Exception, I think the javascript code which handle response would have some error.

**Comment by carlitorweb** [comment]
> The checkAccess method should throw NotAllowed exception if user does not have permission to perform the action like we do for all other extensions (not perform redirection like in proposed code)

Throw a error page it should not be the result for a user. Redirect with a message of why cant access I think is how should work in this case.

**Comment by joomdonation** [comment]
> Throw a error page it should not be the result for a user. Redirect with a message of why cant access I think is how should work in this case.

Users should not be presented with the link if they do not have permission. If they type the URL directly, they get error is expected behavior. Anyway, that's how it works for all component dispatchers which we implemented.

**Comment by carlitorweb** [comment]
Btw, good catch about "sendtestmail"

**Comment by carlitorweb** [comment]
@joomdonation is done

File: administrator/components/com_config/src/Dispatcher/Dispatcher.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// phpcs:disable PSR1.Files.SideEffects
+\defined('_JEXEC') or die;
+// phpcs:enable PSR1.Files.SideEffects
+
+/**
+ * ComponentDispatcher class for com_config
+ *
+ * @since  __DEPLOY_VERSION__
+ */
+class Dispatcher extends \Joomla\CMS\Dispatcher\ComponentDispatcher
```
Comment (joomdonation) [inline administrator/components/com_config/src/Dispatcher/Dispatcher.php]: Could you please add a use statement for `Joomla\CMS\Dispatcher\ComponentDispatcher` and change code to 
`class Dispatcher extends ComponentDispatcher` ? Same for `Joomla\CMS\Access\Exception\NotAllowed` class.

I'm unsure if it is required by our code style rules but we should have it consistent with other component dispatcher, for example https://github.com/joomla/joomla-cms/blob/4.2-dev/administrator/components/com_categories/src/Dispatcher/Dispatcher.php#L25

Comment (carlitorweb) [reply administrator/components/com_config/src/Dispatcher/Dispatcher.php]: I looked for this some weeks ago, but I did not found any code style required for this. I will change it

File: administrator/components/com_config/src/Helper/ConfigHelper.php [hunk 2]
```diff
@@ -73,15 +73,15 @@ public static function hasComponentConfig($component)
...
         $components = self::getAllComponents();
-        $user = Factory::getUser();
+        $user       = Factory::getUser();
 
         // Remove com_config from the array as that may have weird side effects
         $components = array_diff($components, ['com_config']);
 
         foreach ($components as $component) {
-            if (self::hasComponentConfig($component) && (!$authCheck || $user->authorise('core.manage', $component))) {
+            if (self::hasComponentConfig($component) && (!$authCheck || $user->authorise('core.options', $component))) {
```
Comment (joomdonation) [inline administrator/components/com_config/src/Helper/ConfigHelper.php]: I think check for **core.options** permission here is not enough. We need to do the same with permission check as how we do in the dispatcher.

Comment (joomdonation) [reply administrator/components/com_config/src/Helper/ConfigHelper.php]: @carlitorweb Could you please look at my previous comment? Would love to get this issue finally fixed.

**Comment by joomdonation** [comment]
@carlitorweb Unfortunately, your last commits messed up the code. The general rule is you are not allowed to change the behavior of a public method (which could be used by other extensions - potential cause backward incompatible change).

Thinking more about the issue, I realized we need to check it more carefully to avoid user bypass permission check by passing the component input in the request. I come up with this code which should be more secure and a bit more clean than the current code https://github.com/joomla/joomla-cms/compare/4.2-dev...joomdonation:joomla-cms:com_config_dispatcher?expand=1

Please take a look at the code and feel free to ask it to your PR.

**Comment by carlitorweb** [comment]
You are correct about b/c.

@joomdonation I will close this PR, please up yours.

**Comment by joomdonation** [comment]
@carlitorweb OK. If you don't want to continue, you can close the PR and I will open a new PR to replace this one. Thanks for trying !

**Pull request closed without merging**